rust/nasl-interpreter/src/built_in_functions/frame_forgery.rs

```diff
@@ -0,0 +1,797 @@
+}
+
+impl Frame {
+    pub fn new() -> Frame {
+        Frame {
+            srchaddr: MacAddr::zero(),
+            dsthaddr: MacAddr::zero(),
+            ethertype: 0,
+            payload: vec![],
+        }
+    }
+
+    pub fn set_srchaddr(&mut self, srchaddr: MacAddr) -> &Frame {
+        self.srchaddr = srchaddr;
+        self
+    }
+    pub fn set_dsthaddr(&mut self, dsthaddr: MacAddr) -> &Frame {
+        self.dsthaddr = dsthaddr;
+        self
+    }
+    pub fn set_ethertype(&mut self, ethertype: u16) -> &Frame {
+        self.ethertype = ethertype;
+        self
+    }
+
+    pub fn set_payload(&mut self, payload: Vec<u8>) -> &Frame {
+        self.payload = payload;
+        self
+    }
+}
+
+impl Default for Frame {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+impl From<Frame> for Vec<u8> {
+    fn from(f: Frame) -> Vec<u8> {
+        let mut raw_frame = vec![];
+        raw_frame.extend(f.dsthaddr.octets());
+        raw_frame.extend(f.srchaddr.octets());
+        raw_frame.extend(f.ethertype.to_ne_bytes());
+        raw_frame.extend(f.payload);
+        raw_frame
+    }
+}
+
+impl From<&Frame> for Vec<u8> {
+    fn from(f: &Frame) -> Vec<u8> {
+        let mut raw_frame = vec![];
+        raw_frame.extend(f.dsthaddr.octets());
+        raw_frame.extend(f.srchaddr.octets());
+        raw_frame.extend(f.ethertype.to_be_bytes());
+        raw_frame.extend(f.payload.clone());
+        raw_frame
+    }
+}
+
+impl TryFrom<&[u8]> for Frame {
+    type Error = FunctionError;
+
+    fn try_from(f: &[u8]) -> Result<Self, Self::Error> {
+        if f.len() < 14 {
+            Err(FunctionError::new(
+                "get_local_mac_address_from_ip",
+                ("valid ip address").into(),
+            ))
+        } else {
+            let mut frame = Frame::new();
+            frame.set_dsthaddr(MacAddr(f[0], f[1], f[2], f[3], f[4], f[5]));
+            frame.set_srchaddr(MacAddr(f[6], f[7], f[8], f[9], f[10], f[11]));
+            frame.set_ethertype(u16::from_be_bytes([f[12], f[13]]));
+            if f.len() >= 15 {
+                frame.set_payload(f[14..].to_vec());
+            }
+            Ok(frame)
+        }
+    }
+}
+
+impl fmt::Display for Frame {
+    // This trait requires `fmt` with this exact signature.
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut s: String = "".to_string();
+        let vector: Vec<u8> = self.into();
+        let mut count = 0;
+
+        for e in vector {
+            s.push_str(&format!("{:02x}", &e));
+            count += 1;
+            if count % 2 == 0 {
+                s.push(' ');
+            }
+            if count % 16 == 0 {
+                s.push('\n');
+            }
+        }
+        write!(f, "{}", s)
+    }
+}
+
+#[derive(Debug)]
+struct ArpHeader {
+    ar_hrd: u16,
+    ar_pro: u16,
+    ar_hln: u8,
+    ar_pln: u8,
+    ar_op: u16,
+}
+
+const ARP_HEADER: ArpHeader = ArpHeader {
+    ar_hrd: ARPHRD_ETHER,
```

Comment:
why do you set a prefix there? isn't it clear that it is `ar_` due to ArpHeader?

Reply:
I just followed the naming from the c implementation.